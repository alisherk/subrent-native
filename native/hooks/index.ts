export * from './usePaginateQuery'; 
export * from './useDocument';
export * from './useCollection';