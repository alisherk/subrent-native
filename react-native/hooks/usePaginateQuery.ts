import { useReducer, useEffect, useRef } from 'react';
import { RentalModel } from 'models';

enum ActionTypes {
  INITIAL_LOAD = 'INITIAL_LOAD',
  LOAD_MORE = 'LOAD_MORE',
  ERROR = 'ERROR',
  RESET = 'RESET', 
  RELOADING = 'RELOADING'
}

type State = {
  hasMore: boolean;
  triggerEffect: firebase.firestore.DocumentData | null;
  limit: number;
  rentals: RentalModel[];
  lastLoaded: firebase.firestore.QueryDocumentSnapshot | null;
  loading: boolean;
  error: null | Error | string;
  hasMounted: boolean;
  reloading: boolean;
};

type Action =
    {
      type: ActionTypes.INITIAL_LOAD;
      snapshot: firebase.firestore.QuerySnapshot;
      limit: number;
    }
  | { type: ActionTypes.LOAD_MORE }
  | { type: ActionTypes.ERROR; error: string }
  | { type: ActionTypes.RESET }
  | { type: ActionTypes.RELOADING }

function reducer(state: State, action: Action): State {
  switch (action.type) {
    case ActionTypes.INITIAL_LOAD: {
      const rentals = [...state.rentals];   
      if (action.snapshot) {
        action.snapshot.forEach((doc) => {
          if (doc.data) {
            const rental = doc.data() as Omit<RentalModel, 'id'> ;
            rentals.push({
              id: doc.id,
              ...rental, 
            });
          }
        });
      }

      const nextLimit = rentals.length + action.limit;
      const end = rentals.length < action.limit || nextLimit === state.limit;

      return {
        ...state,
        rentals,
        hasMore: !end,
        limit: nextLimit,
        loading: false,
        lastLoaded:
          action.snapshot.docs[action.snapshot.docs.length - 1] ||
          state.lastLoaded,
      };
    }

    case ActionTypes.LOAD_MORE: {
      if (state.hasMore) {
        return {
          ...state,
          triggerEffect: state.lastLoaded,
        };
      }
      return state;
    }
    case ActionTypes.ERROR:
      return {
        ...state,
        error: action.error,
        loading: false,
      };
     case ActionTypes.RELOADING: 
      return {
          ...state, 
          reloading: true
      }
     case ActionTypes.RESET: 
      return {
          ...state, 
          reloading: false, 
          lastLoaded: null, 
          rentals: [],
          triggerEffect: null
      }
    default:
      return state;
  }
}

const initialState: State = {
  hasMore: false,
  triggerEffect: null,
  limit: 0,
  rentals: [],
  lastLoaded: null,
  loading: true,
  error: null,
  hasMounted: false,
  reloading: false
};

export type PaginateHookData = {
  loadMore: () => void;
  resetLoad: () => void;
  error: null | Error | string;
  loading: boolean;
  reloading: boolean;
  hasMore: boolean;
  rentals: RentalModel[];
};
export const usePaginateQuery = (
  query: firebase.firestore.Query,
  { limit = 1 } = {}
): PaginateHookData => {
  const [state, dispatch] = useReducer(reducer, initialState);

  const { lastLoaded, hasMore, triggerEffect, error, rentals, loading, reloading } = state;

  const persisted = useRef<firebase.firestore.Query>(query);
  const lastVisible = useRef<firebase.firestore.DocumentData | null>(lastLoaded);

  useEffect(() => {
    lastVisible.current = lastLoaded;
  }, [lastLoaded]);

  useEffect(() => {
    let isCancelled: boolean = false;
    async function getData(): Promise<void> {
      let query;
      lastVisible.current
        ? (query = persisted.current
            .startAfter(lastVisible.current)
            .limit(limit))
        : (query = persisted.current.limit(limit));
      try {
        const snapshot = await query.get();
        if (!isCancelled) {
          dispatch({ type: ActionTypes.INITIAL_LOAD, snapshot, limit });
        }
      } catch (err) {
        dispatch({ type: ActionTypes.ERROR, error: err.message });
      }
    }
    getData();
    return () => {
      isCancelled = true;
    };
  }, [triggerEffect, limit]);

  function loadMore() {
    dispatch({ type: ActionTypes.LOAD_MORE });
  }

  function resetLoad() {
      dispatch({ type: ActionTypes.RELOADING}); 
      dispatch({ type: ActionTypes.RESET })
  }

  return {
    loadMore,
    resetLoad,
    error,
    loading,
    hasMore,
    rentals,
    reloading
  };
};
