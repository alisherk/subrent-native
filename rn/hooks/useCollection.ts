import { useState, useEffect, useRef, useCallback } from 'react';
import { firestore } from 'firebase';
import { Rental } from 'common';

type Data = Rental[];
type CollectionHook = [Data, boolean, Error | null];

export const useCollection = (
  query: firestore.Query,
  deps?: string[]
): CollectionHook => {
  const [data, setData] = useState<Data>([]);
  const [loading, setLoading] = useState(false);
  const [err, setErr] = useState(null);
  const ref = useRef(query);

  const fetchData = useCallback(async () => {
    try {
      setLoading(true);
      const snap = await ref.current.get();
      if (snap.empty) {
        setLoading(false);
        return;
      }
      const results = snap.docs.map((doc) => {
        const rental = doc.data() as Omit<Rental, 'id'>; 
        return {
          id: doc.id,
          ...rental
        };
      });
      setData(results);
      setLoading(false);
    } catch (err) {
      setErr(err.message);
    }
  }, []);

  useEffect(() => {
    if (!ref.current) return;
    fetchData();
  }, [fetchData, deps]);

  return [data, loading, err];
};
