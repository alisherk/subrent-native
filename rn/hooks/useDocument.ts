import { useReducer, useEffect, useRef } from 'react';
import { RentalModel } from 'types';
import type firebase from 'firebase';

enum ActionTypes {
  ON_FETCH = 'ON_FETCH',
  ON_ERROR = 'ON_ERROR',
  ON_LOAD = 'ON_LOAD',
  RELOAD = 'RELOAD',
}

type Status = 'loading' | 'loaded' | 'error' | null;

type State = {
  data: RentalModel | null;
  status: Status;
  error: string | null;
  stateUpdate: number;
};

type Action =
  | {
      type: ActionTypes.ON_FETCH;
      document: firebase.firestore.DocumentData;
    }
  | { type: ActionTypes.ON_ERROR; error: string | null }
  | { type: ActionTypes.ON_LOAD }
  | { type: ActionTypes.RELOAD };

function reducer(state: State, action: Action): State {
  switch (action.type) {
    case ActionTypes.RELOAD:
      return {
        ...state,
        stateUpdate: state.stateUpdate + 1,
      };
    case ActionTypes.ON_LOAD:
      return {
        ...state,
        status: 'loading',
      };
    case ActionTypes.ON_FETCH:
      return {
        ...state,
        status: 'loaded',
        data: {
          id: action.document.id,
          ...action.document.data(),
        },
      };
    case ActionTypes.ON_ERROR:
      return {
        ...state,
        error: action.error,
        status: 'error',
      };
    default:
      return state;
  }
}

const initialState: State = {
  data: null,
  status: null,
  error: null,
  stateUpdate: 0,
};

type reload = () => void;

export type useDocumentHookData = [
  RentalModel | null,
  Status,
  string | null,
  reload
];

export const useDocument = (
  docRef: firebase.firestore.DocumentReference | null,
  deps: any[]
): useDocumentHookData => {
  const [state, dispatch] = useReducer(reducer, initialState);
  const ref = useRef(docRef);

  const { data, status, error, stateUpdate } = state;

  const reload = (): void => dispatch({ type: ActionTypes.RELOAD });

  useEffect(() => {
    if (!ref.current) return;
    (async function fetchData() {
      try {
        dispatch({ type: ActionTypes.ON_LOAD });
          const document = await ref.current?.get();
          if (!document?.exists) {
            return dispatch({
              type: ActionTypes.ON_ERROR,
              error: 'Document is empty',
            });
          }
          dispatch({ type: ActionTypes.ON_FETCH, document });
      } catch (err) {
        dispatch({ type: ActionTypes.ON_ERROR, error: err.message });
      }
    })();
  
  }, [stateUpdate, ...deps]);

  return [data, status, error, reload];
};
