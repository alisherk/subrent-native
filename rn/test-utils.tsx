import React from 'react';
import { render } from '@testing-library/react-native';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux';
import { rootReducer } from './redux/reducers';
import { PaginateHookData } from './hooks';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

//mock redux store cofig for dispathing action 
const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
export const returnWithStore = (args?: any) =>  mockStore(args);

//testing libriary config for returning render with redux
export function renderWithRedux(
    component: React.ReactNode,
    initialState?: any,
    {
      store = createStore(rootReducer, initialState, applyMiddleware(thunk)),
    } = {}
  ) {
    return {
      ...render(<Provider store={store}>{component}</Provider>),
      store,
    };
  }
  
  //function for passing props to components with react navigation
  export const createTestProps =<T extends {}> (props: T) => ({
    navigation: {
      navigate: jest.fn(),
      setOptions: jest.fn(), 
      addListener: jest.fn()
    },
    route: {
      params: {}
    },
    ...props,
  });

//mock data used in tests
export const mockRental = {
  id: '1',
  name: 'test',
  full_day_price: '10',
  half_day_price: '5',
  category: 'test',
  region: 'test',
  description: 'test',
  ownerImage: 'test',
  displayName: 'test',
  delivery: 'test',
  image: 'test',
  instructions: 'test', 
  confirmation_required: 'test',
  ownerUid: 'e9e9e9er', 
  expoToken: 'ajdjdudnfififh',
  g: { geopoint: { U: 1, k: 2 }}
};

export const mockPaginateHookReturn: PaginateHookData = {
  rentals: [mockRental], 
  loadMore: jest.fn, 
  resetLoad: jest.fn, 
  loading: false, 
  reloading: false, 
  error: null, 
  hasMore: true, 
}


