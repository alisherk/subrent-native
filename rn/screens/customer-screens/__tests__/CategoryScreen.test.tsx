import React from 'react';
import { CategoryScreen } from '..';
import { renderWithRedux, createTestProps, mockPaginateHookReturn } from 'test-utils';
import * as hooks from 'hooks/usePaginateQuery';

it('renders without crashing', () => {
  jest.spyOn(hooks, 'usePaginateQuery').mockReturnValue(mockPaginateHookReturn);
  const props: any = createTestProps({});
  const { toJSON } = renderWithRedux(<CategoryScreen {...props} />);
  expect(toJSON()).toMatchSnapshot(); 
});
