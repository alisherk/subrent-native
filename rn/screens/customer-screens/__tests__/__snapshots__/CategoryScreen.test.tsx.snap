// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<RCTScrollView
  data={
    Array [
      Object {
        "category": "test",
        "confirmation_required": "test",
        "delivery": "test",
        "description": "test",
        "displayName": "test",
        "full_day_price": "10",
        "half_day_price": "5",
        "id": "1",
        "image": "test",
        "instructions": "test",
        "name": "test",
        "ownerImage": "test",
        "region": "test",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={2}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReached={[Function]}
  onEndReachedThreshold={0.2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onRefresh={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  refreshControl={
    <RefreshControlMock
      onRefresh={[Function]}
      refreshing={false}
    />
  }
  refreshing={false}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <RCTRefreshControl />
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <RCTSafeAreaView
        emulateUnlessSupported={true}
        style={
          Object {
            "backgroundColor": "transparent",
            "flex": 1,
          }
        }
      >
        <RCTScrollView
          automaticallyAdjustContentInsets={false}
          contentContainerStyle={
            Array [
              Object {
                "padding": 10,
              },
              undefined,
            ]
          }
          contentInset={
            Object {
              "bottom": 0,
            }
          }
          enableAutomaticScroll={true}
          enableOnAndroid={false}
          enableResetScrollToCoords={true}
          extraHeight={75}
          extraScrollHeight={0}
          getScrollResponder={[Function]}
          handleOnScroll={[Function]}
          keyboardDismissMode="interactive"
          keyboardOpeningTime={250}
          keyboardShouldPersistTaps="handled"
          keyboardSpace={0}
          onScroll={[Function]}
          padder={true}
          resetKeyboardSpace={[Function]}
          resetScrollToCoords={
            Object {
              "x": 0,
              "y": 0,
            }
          }
          scrollEventThrottle={1}
          scrollForExtraHeightOnAndroid={[Function]}
          scrollIntoView={[Function]}
          scrollToEnd={[Function]}
          scrollToFocusedInput={[Function]}
          scrollToPosition={[Function]}
          showsVerticalScrollIndicator={true}
          style={
            Object {
              "backgroundColor": "transparent",
              "flex": 1,
            }
          }
          update={[Function]}
          viewIsInsideTabBar={false}
        >
          <View>
            <View
              style={
                Object {
                  "backgroundColor": "#fff",
                  "borderBottomWidth": 0.5,
                  "borderColor": "#ccc",
                  "borderLeftWidth": 0.5,
                  "borderRadius": 2,
                  "borderRightWidth": 0.5,
                  "borderTopWidth": 0.5,
                  "elevation": 3,
                  "flexWrap": "nowrap",
                  "marginBottom": 5,
                  "marginLeft": 2,
                  "marginRight": 2,
                  "marginTop": 5,
                  "shadowColor": "#000",
                  "shadowOffset": Object {
                    "height": 2,
                    "width": 0,
                  },
                  "shadowOpacity": 0.1,
                  "shadowRadius": 1.5,
                }
              }
            >
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                <View
                  bordered={true}
                  style={
                    Array [
                      Object {
                        "alignItems": "center",
                        "backgroundColor": "#fff",
                        "borderBottomWidth": 0.5,
                        "borderColor": "#ccc",
                        "borderRadius": 2,
                        "flexDirection": "row",
                        "paddingBottom": 10,
                        "paddingLeft": 15,
                        "paddingRight": 15,
                        "paddingTop": 10,
                      },
                      Object {
                        "justifyContent": "center",
                      },
                    ]
                  }
                >
                  <Text
                    style={
                      Object {
                        "color": "#000",
                        "fontFamily": "System",
                        "fontSize": 16,
                      }
                    }
                  >
                     
                    test
                     
                  </Text>
                </View>
                <View
                  cardBody={true}
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#fff",
                      "borderRadius": 2,
                      "flexDirection": "row",
                      "paddingBottom": -5,
                      "paddingLeft": -5,
                      "paddingRight": -5,
                      "paddingTop": -5,
                    }
                  }
                >
                  <Image
                    source={
                      Object {
                        "uri": "test",
                      }
                    }
                    style={
                      Object {
                        "flex": 1,
                        "height": 250,
                      }
                    }
                  />
                </View>
                <View
                  bordered={true}
                  style={
                    Array [
                      Object {
                        "alignItems": "center",
                        "backgroundColor": "#fff",
                        "borderBottomWidth": 0.5,
                        "borderColor": "#ccc",
                        "borderRadius": 2,
                        "flexDirection": "row",
                        "paddingBottom": 10,
                        "paddingLeft": 15,
                        "paddingRight": 15,
                        "paddingTop": 10,
                      },
                      Object {
                        "justifyContent": "space-evenly",
                      },
                    ]
                  }
                >
                  <Text
                    style={
                      Object {
                        "color": "#000",
                        "fontFamily": "System",
                        "fontSize": 16,
                      }
                    }
                  >
                     Full day: $
                    10
                     
                  </Text>
                  <Text
                    style={
                      Object {
                        "color": "#000",
                        "fontFamily": "System",
                        "fontSize": 16,
                      }
                    }
                  >
                     Half day: $
                    5
                     
                  </Text>
                </View>
                <View
                  style={
                    Array [
                      Object {},
                      Object {
                        "padding": 10,
                      },
                    ]
                  }
                >
                  <View
                    accessible={true}
                    focusable={true}
                    forwardedRef={[Function]}
                    onClick={[Function]}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Object {
                        "alignItems": "center",
                        "alignSelf": "stretch",
                        "backgroundColor": "#5cb85c",
                        "borderBottomWidth": null,
                        "borderColor": "#007aff",
                        "borderLeftWidth": null,
                        "borderRadius": 0,
                        "borderRightWidth": null,
                        "borderTopWidth": null,
                        "borderWidth": undefined,
                        "elevation": 2,
                        "flexDirection": "row",
                        "height": 45,
                        "justifyContent": "center",
                        "opacity": 1,
                        "paddingBottom": 6,
                        "paddingTop": 6,
                        "shadowColor": undefined,
                        "shadowOffset": undefined,
                        "shadowOpacity": undefined,
                        "shadowRadius": undefined,
                      }
                    }
                  >
                    <Text
                      style={
                        Object {
                          "backgroundColor": "transparent",
                          "color": "#fff",
                          "fontFamily": "System",
                          "fontSize": 16.5,
                          "marginLeft": 0,
                          "marginRight": 0,
                          "paddingLeft": 16,
                          "paddingRight": 16,
                        }
                      }
                    >
                      View
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </RCTScrollView>
      </RCTSafeAreaView>
    </View>
  </View>
</RCTScrollView>
`;
